### PR TITLE
feat(FEC-8452): playlist

### PR DESCRIPTION
**Pull request by yairans**
### Description of the Changes

add playlist api: 
* `loadPlaylist` - to load playlist from the backend
* `loadPlaylistByEntryList` - to load playlist by entry id's
* `setPlaylist`
* `getter playlist` - for the playlist api (`playNext`, `playPrev` etc.)

new classes: `PlaylistManager`, `Playlist` and `PlaylistItem`

### CheckLists

- [x] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [ ] test are passing in local environment 
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [ ] Docs have been updated

### CODE REVIEW

File: src/common/playlist/playlist-manager.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    );
+  }
+
+  _setActiveItem(autoplay: ?boolean) {
+    const activeItem = this._config.items[this._activeItemIndex];
+    this._player.dispatchEvent(new FakeEvent(PlaylistEventType.PLAYLIST_ITEM_CHANGED, {index: this._activeItemIndex, activeItem}));
+    if (typeof autoplay === 'boolean') {
+      this._player.configure({playback: {autoplay}});
+    }
+    if (activeItem.sources.hls.length) {
```
Comment (OrenMe) [inline src/common/playlist/playlist-manager.js]: better have another condition for this, hls is not enough, as u can have only dash, progressive etc. use utility method to check if you have any of those.

Comment (OrenMe) [reply src/common/playlist/playlist-manager.js]: consider making this method always return a promise for future handling of flow.
Currently you have one branch returning sync, where you ahve the sources data, and one async, when you load the media data from API.

File: src/common/playlist/playlist-manager.js [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    } else {
+      this._player.loadMedia({entryId: activeItem.sources.id}).then(mediaConfig => {
+        Utils.Object.mergeDeep(activeItem.sources, mediaConfig.sources);
+      });
+    }
+  }
+
+  getActiveSources(): Promise<*> {
+    const activeItem = this._config.items[this._activeItemIndex];
+    if (activeItem.sources.hls.length) {
```
Comment (OrenMe) [inline src/common/playlist/playlist-manager.js]: again, utility method

File: src/common/playlist/playlist-manager.js [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  _setActiveItem(autoplay: ?boolean) {
+    const activeItem = this._config.items[this._activeItemIndex];
+    this._player.dispatchEvent(new FakeEvent(PlaylistEventType.PLAYLIST_ITEM_CHANGED, {index: this._activeItemIndex, activeItem}));
+    if (typeof autoplay === 'boolean') {
+      this._player.configure({playback: {autoplay}});
+    }
+    if (activeItem.sources.hls.length) {
+      // reach the condition
+      this._player.reset();
+      this._player.configure(activeItem); // maybe use setMedia but need session and plugins
```
Comment (OrenMe) [inline src/common/playlist/playlist-manager.js]: your comment is correct, setMedia does additional settings that you must have

File: src/common/playlist/playlist-manager.js [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    } else {
+      this._player.loadMedia({entryId: activeItem.sources.id}).then(mediaConfig => {
+        Utils.Object.mergeDeep(activeItem.sources, mediaConfig.sources);
+      });
+    }
+  }
+
+  getActiveSources(): Promise<*> {
```
Comment (OrenMe) [inline src/common/playlist/playlist-manager.js]: why do we need this API?

File: src/common/playlist/playlist-manager.js [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    const activeItem = this._config.items[this._activeItemIndex];
+    if (activeItem.sources.hls.length) {
+      return Promise.resolve(activeItem.sources);
+    } else {
+      this._player.loadMedia({entryId: activeItem.sources.id}).then();
```
Comment (OrenMe) [inline src/common/playlist/playlist-manager.js]: empty `then`?

File: src/common/playlist/playlist-manager.js [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    return (
+      (activeItem.countdown && activeItem.countdown.timeToShow) ||
+      this._player.duration - ((activeItem.countdown && activeItem.countdown.duration) || this._config.countdown.duration)
+    );
+  }
+
+  _setActiveItem(autoplay: ?boolean) {
+    const activeItem = this._config.items[this._activeItemIndex];
+    this._player.dispatchEvent(new FakeEvent(PlaylistEventType.PLAYLIST_ITEM_CHANGED, {index: this._activeItemIndex, activeItem}));
+    if (typeof autoplay === 'boolean') {
```
Comment (OrenMe) [inline src/common/playlist/playlist-manager.js]: manage this in the API that uses playNext/Prev and just call `play`.
If you combine it with the comment below to make this method return a promise you can also add error handling.

File: src/common/playlist/playlist-manager.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+  getActiveSources(): Promise<*> {
+    const activeItem = this._config.items[this._activeItemIndex];
+    if (activeItem.sources.hls.length) {
+      return Promise.resolve(activeItem.sources);
+    } else {
+      this._player.loadMedia({entryId: activeItem.sources.id}).then();
+    }
+  }
+
+  playNext() {
```
Comment (OrenMe) [inline src/common/playlist/playlist-manager.js]: add log

File: src/common/playlist/playlist-manager.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    if (typeof this._activeItemIndex === 'number' && this._config.items[this._activeItemIndex + 1]) {
+      this._activeItemIndex++;
+      this._setActiveItem(true);
+    }
+  }
+
+  playPrev() {
+    if (typeof this._activeItemIndex === 'number' && this._config.items[this._activeItemIndex - 1]) {
```
Comment (OrenMe) [inline src/common/playlist/playlist-manager.js]: add log

Comment (OrenMe) [inline src/common/playlist/playlist-manager.js]: use a method to validate that index is number and is in range, as you repeat it in this class

File: src/kaltura-player.js [hunk 9]
```diff
@@ -54,6 +53,7 @@ class KalturaPlayer extends FakeEventTarget {
 
   setMedia(mediaConfig: ProviderMediaConfigObject): void {
     this._logger.debug('setMedia', mediaConfig);
+    this.reset();
```
Comment (OrenMe) [inline src/kaltura-player.js]: can't move this here - expectiation is that once loadMedia request is done the player will reset itself immediately and not keep playing until API server responds.
If you call this API via the playlist manager then it should call reset by itself.

File: src/common/playlist/playlist-manager.js [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    return (
+      (activeItem.countdown && activeItem.countdown.timeToShow) ||
+      this._player.duration - ((activeItem.countdown && activeItem.countdown.duration) || this._config.countdown.duration)
+    );
+  }
+
+  _setActiveItem(): Promise<*> {
+    const activeItem = this._config.items[this._activeItemIndex];
+    const prevItem = this._config.items[this._activeItemIndex - 1];
+    const nextItem = this._config.items[this._activeItemIndex + 1];
```
Comment (itaykinnrot) [inline src/common/playlist/playlist-manager.js]: @yairans need to add % or check the length

Comment (OrenMe) [reply src/common/playlist/playlist-manager.js]: @itaykinnrot the pointers will just be undefined in this case and I think that's OK, cause we don't want cyclic next/prev on first/last items in a list.
Think of it like this - on first entry of playlist on the first time you will have a back button control that shows the last item of the playlist.

File: src/common/playlist/playlist-manager.js [hunk 11]
```diff
@@ -58,53 +52,74 @@ class PlaylistManager {
     );
   }
 
-  _setActiveItem(autoplay: ?boolean) {
+  _setActiveItem(): Promise<*> {
     const activeItem = this._config.items[this._activeItemIndex];
+    const prevItem = this._config.items[this._activeItemIndex - 1];
+    const nextItem = this._config.items[this._activeItemIndex + 1];
+    PlaylistManager._logger.debug(`Playing item number ${this._activeItemIndex}`, activeItem);
     this._player.dispatchEvent(new FakeEvent(PlaylistEventType.PLAYLIST_ITEM_CHANGED, {index: this._activeItemIndex, activeItem}));
```
[the remark this answers is not in the document]
Comment (yairans) [reply src/common/playlist/playlist-manager.js]: removed

File: src/common/playlist/playlist-manager.js [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+    this.configure(options.playlist);
+    this.addBindings();
+  }
+
+  configure(config: ?KPPlaylistConfigObject) {
+    if (config) {
+      this._playlist.configure(config);
+      Utils.Object.mergeDeep(this._options, config.options);
+      Utils.Object.mergeDeep(this._countdown, config.countdown);
+      if (this._playlist.items.find(item => !!item.sources)) {
```
Comment (dan-ziv) [inline src/common/playlist/playlist-manager.js]: Maybe create `this._playlist.hasPlayableItems()` API for the playlist and do that logic inside the playlist itself?

File: src/common/playlist/playlist.js [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+    this._activeItemIndex = -1;
+  }
+
+  configure(config: KPPlaylistConfigObject) {
+    this._id = config.id;
+    Utils.Object.mergeDeep(this._metadata, config.metadata);
+    if (config.items) {
+      this._items = [];
+      config.items.forEach(item => {
+        this._items.push(new PlaylistItem(item.sources, item.config));
```
Comment (dan-ziv) [inline src/common/playlist/playlist.js]: Why not to send the item itself in the constructor and extract the necessary data from it inside the `PlaylistItem`?

File: src/kaltura-player.js [hunk 14]
```diff
@@ -38,9 +40,10 @@ class KalturaPlayer extends FakeEventTarget {
     super();
     this._eventManager = new EventManager();
     this._localPlayer = loadPlayer(options);
+    this._logger = getLogger('KalturaPlayer' + Utils.Generator.uniqueId(5));
     this._uiWrapper = new UIWrapper(this, options);
     this._provider = new Provider(options.provider, __VERSION__);
-    this._logger = getLogger('KalturaPlayer' + Utils.Generator.uniqueId(5));
+    this._playlistManager = new PlaylistManager(this, options);
```
Comment (dan-ziv) [inline src/kaltura-player.js]: Are we creating the playlist manager anyway?

**Pull request merged**